[14.0] [IMP] connector_search_engine: Use real recompute_json batches with dichotomic retry on fail

paradoxxxzero opened the pull request:
This PR batches the `recompute_json` jobs together instead of batching the job creation.
It also adds a dichotomic retry on failing batches, which is nice to quickly isolate failing bindings.

Superseeds #100

@sebastienbeau

hparfr commented:
@paradoxxxzero can you please fix this PR ?

simahawk commented:
Is this work finished?
Also, I wonder if we shouldn't leverage [`identity_key`](https://github.com/OCA/queue/blob/92eae4b427578068c5d9256d51bf672938bc9422/queue_job/models/base.py#L64) to avoid duplication of jobs.

sebastienbeau commented on connector_search_engine/models/se_binding.py:
```diff
@@ -105,14 +107,59 @@ def _msg_cannot_delete_not_synchronized(self):
 
+    def recompute_json(self, force_export=False, retry=True):
+        try:
+            return self._recompute_json(force_export=force_export)
```
We need to add a savepoint here to be able to rollback in case of exception
```
try:
    with env.cr.savepoint():
        return self._recompute_json(force_export=force_export)
exception Exception:
    ....

```

sebastienbeau commented on connector_search_engine/models/se_binding.py:
```diff
@@ -105,14 +107,59 @@ def _msg_cannot_delete_not_synchronized(self):
+                    force_export=force_export,
+                    batch_size=math.ceil(len(self) / 2),
+                    retry=True,
+                )
```
In case that we split the job, it will be great to return something here

return "Job have been splited due to failling element"

So this will be set in the result of the job in the UI

sebastienbeau commented on connector_search_engine/models/se_binding.py:
```diff
@@ -105,14 +107,59 @@ def _msg_cannot_delete_not_synchronized(self):
+            return self._recompute_json(force_export=force_export)
+        except Exception:
+            # If the batch fails, retry with a half len batch:
+            if retry and len(self) > 1:
```
I do not think that this new keywork "retry=True" is needed (we can remove it from function args).
If we want to really want to inactive/customize this feature in custom module, I would prefer to add a method

```
def _should_be_retry(self):
    return len(self) >1
...
...
...
if self._should_be_retry(sefl):
    ...

``

sebastienbeau commented:
@paradoxxxzero, great work, minor comment to change
We should merge it next week

paradoxxxzero commented:
@sebastienbeau I merged your changes and rebased

sebastienbeau reviewed (approved):
LGTM (code review) and test

sebastienbeau commented:
@simahawk used in production, are you ok for merging it ?

simahawk commented on connector_search_engine/models/se_binding.py:
```diff
@@ -103,14 +105,63 @@ def unlink(self):
+
+            # We check if we are currently handling an exception in order
+            # to change the job description.
+            (current_exception, current_exception_value, __) = sys.exc_info()
```
line 119 to 141 can be moved to `_jobify_get_job_description` or alike to make the whole thing more readable

simahawk commented on connector_search_engine/models/se_binding.py:
```diff
@@ -103,14 +105,63 @@ def unlink(self):
+                    force_export=force_export,
+                    batch_size=math.ceil(len(self) / 2),
+                )
+                return _(
+                    "Job have been splited due to failling element.\nError: {}"
+                ).format(e)
```
```suggestion
                return _(
                    "Jobs have been split due to failing element.\nError: %s"
                ) % str(e)
```

simahawk commented on connector_search_engine/models/se_binding.py:
```diff
@@ -103,14 +105,63 @@ def unlink(self):
+            description = (
+                (
+                    _(
+                        "Batch task of %s for recompute %s json after previous "
+                        "batch task raised %s: %s."
```
```suggestion
                        "Batch of %s items for %s. Previous batch raised %s: %s."
```
The fact that we are recomputing json is in the job func name already. Let's keep this msg shorter ;)

sebastienbeau commented:
@simahawk done ! thanks for your feedback

sebastienbeau commented:
@simahawk can you review?

simahawk commented:
I'll have a look today :sweat_smile:

simahawk commented on connector_search_engine/models/se_binding.py:
```diff
@@ -103,14 +105,60 @@ def unlink(self):
+                    batch_size=math.ceil(len(self) / 2),
+                )
+                return _(
+                    "Job have been splited due to failling element.\nError: {}"
```
```suggestion
                    "Job have been split because of failure.\nError: {}"
```

simahawk commented on connector_search_engine/models/se_binding.py:
```diff
@@ -103,14 +105,60 @@ def unlink(self):
+        bindings = self.with_context(tracking_disable=True)
+        while bindings:
+            processing, bindings = self._get_binding_to_process(bindings, batch_size)
+            description = self._jobify_get_job_description(len(processing))
```
```suggestion
            description = self._jobify_get_job_description(processing)
```
and get the count inside that method

simahawk reviewed (commented):
@paradoxxxzero thanks looks very good. 
I have 2 more remarks and and plea for the commit: can you shorten it and add a description?

```
[IMP] connector_search_engine: dichotomic retry on fail

When a job containing N bindings fails
the recordset of bindings will be split by N/2
and new jobs will be scheduled.
```
Thanks :pray:

paradoxxxzero committed:
[IMP] connector_search_engine: dichotomic retry on fail

When a job containing N bindings fails
the recordset of bindings will be split by N/2
and new jobs will be scheduled.

simahawk commented:
We need this to fix pre-commit https://github.com/OCA/search-engine/pull/137

simahawk commented:
@paradoxxxzero can you rebase?

simahawk commented:
/ocabot merge minor
